In [1]:
import numpy as np
import pandas as pd




**What is lable encoding**
- convert_categorical to numerical called(Label encoding)


**why lable encoding**

-In order to apply ML algorithm, you need to process only numerical values because all ML alogorithm developed by maths


In [50]:
bank_data=pd.read_csv('/Users/sheelendersharma1/Desktop/data sciece/bank.csv',sep=';')

In [6]:
bank_data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [12]:
bank_data.describe()

age       balance          day     duration     campaign  \
count  4521.000000   4521.000000  4521.000000  4521.000000  4521.000000   
mean     41.170095   1422.657819    15.915284   263.961292     2.793630   
std      10.576211   3009.638142     8.247667   259.856633     3.109807   
min      19.000000  -3313.000000     1.000000     4.000000     1.000000   
25%      33.000000     69.000000     9.000000   104.000000     1.000000   
50%      39.000000    444.000000    16.000000   185.000000     2.000000   
75%      49.000000   1480.000000    21.000000   329.000000     3.000000   
max      87.000000  71188.000000    31.000000  3025.000000    50.000000   

             pdays     previous  
count  4521.000000  4521.000000  
mean     39.766645     0.542579  
std     100.121124     1.693562  
min      -1.000000     0.000000  
25%      -1.000000     0.000000  
50%      -1.000000     0.000000  
75%      -1.000000     0.000000  
max     871.000000    25.000000

In [17]:
cati=bank_data.columns[bank_data.dtypes=='O']


In [18]:
cati

Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact',
       'month', 'poutcome', 'y'],
      dtype='object')

In [20]:
nums=bank_data.columns[(bank_data.dtypes=='int64')]

Index(['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous'], dtype='object')

In [ ]:
## Method-1

#Manual conversion categorical to numerical

In [22]:
bank_data['loan'].value_counts()

no     3830
yes     691
Name: loan, dtype: int64

In [23]:
bank_data['loan'].unique()
#  There are two fields are there
# one is yes anf another is no

array(['no', 'yes'], dtype=object)

In [26]:
# create a dict 
dict1={"no":0,"yes":1}
bank_data['loan_new']=bank_data['loan'].map(dict1)

In [28]:
bank_data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  \
0  cellular   19   oct        79         1     -1         0  unknown  no   
1  cellular   11   may       220         1    339         4  failure  no   
2  cellular   16   apr       185         1    330         1  failure  no   
3   unknown    3   jun       199         4     -1         0  unknown  no   
4   unknown    5   may       226         1     -1         0  unknown  no   

   loan_new  
0         0  
1         1  
2         0  
3         1  
4         0

In [ ]:
# I am droping two coloumns
 # 1. Loan_new,2.number

In [29]:
bank_data.drop("loan_new",axis=1,inplace=True)

In [30]:
bank_data.drop("",axis=1,inplace=True)

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

In [32]:
# create a dict 
dict1={"no":0,"yes":1}
bank_data['loan']=bank_data['loan'].map(dict1)

In [33]:
bank_data.head()

age          job  marital  education default  balance housing  loan  \
0   30   unemployed  married    primary      no     1787      no     0   
1   33     services  married  secondary      no     4789     yes     1   
2   35   management   single   tertiary      no     1350     yes     0   
3   30   management  married   tertiary      no     1476     yes     1   
4   59  blue-collar  married  secondary      no        0     yes     0   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [51]:
dict1={"married":1,"single":0}
bank_data['marital']=bank_data['marital'].map(dict1)

In [52]:
bank_data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed      1.0    primary      no     1787      no   no   
1   33     services      1.0  secondary      no     4789     yes  yes   
2   35   management      0.0   tertiary      no     1350     yes   no   
3   30   management      1.0   tertiary      no     1476     yes  yes   
4   59  blue-collar      1.0  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [55]:
bank_data=pd.read_csv('/Users/sheelendersharma1/Desktop/data sciece/bank.csv',sep=';')
bank_data["marital"].value_counts()
bank_data["marital"].unique()
dict2={"married":0,"single":1,"divorced":2}
bank_data["marital_new"]=bank_data["marital"].map(dict2) # Don't use this method if lables are more than 3

In [56]:
bank_data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  \
0  cellular   19   oct        79         1     -1         0  unknown  no   
1  cellular   11   may       220         1    339         4  failure  no   
2  cellular   16   apr       185         1    330         1  failure  no   
3   unknown    3   jun       199         4     -1         0  unknown  no   
4   unknown    5   may       226         1     -1         0  unknown  no   

   marital_new  
0            0  
1            0  
2            1  
3            0  
4            0

## method -2 
**using numpy**

In [57]:
bank_data=pd.read_csv('/Users/sheelendersharma1/Desktop/data sciece/bank.csv',sep=';')

In [58]:
bank_data

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

**np.where**

In [59]:
condition=bank_data['loan']=='no'
# whenever this condition is true replace with 0 i.e no is replace by 0
# whenever the condition is not True replace with 1 i.e 
bank_data['loan']=np.where(condition,0,1)

In [60]:
bank_data['loan'].value_counts()

0    3830
1     691
Name: loan, dtype: int64

# Method:3

**np.dummies**

In [ ]:
# suppose Marital
 Marital
1)single
2)divorce
3)Married


# it will create three more variables
 Marital_single     Marital_Divorce      Marital_married
    1                  0                      0
    0                  1                      1
    
    
    
    

In [62]:
new_data=pd.get_dummies(bank_data,columns=['marital'],prefix='marital')

# orthogonal ----90degree
# one is other will off

In [63]:
new_data

age            job  education default  balance housing  loan   contact  \
0      30     unemployed    primary      no     1787      no     0  cellular   
1      33       services  secondary      no     4789     yes     1  cellular   
2      35     management   tertiary      no     1350     yes     0  cellular   
3      30     management   tertiary      no     1476     yes     1   unknown   
4      59    blue-collar  secondary      no        0     yes     0   unknown   
...   ...            ...        ...     ...      ...     ...   ...       ...   
4516   33       services  secondary      no     -333     yes     0  cellular   
4517   57  self-employed   tertiary     yes    -3313     yes     1   unknown   
4518   57     technician  secondary      no      295      no     0  cellular   
4519   28    blue-collar  secondary      no     1137      no     0  cellular   
4520   44   entrepreneur   tertiary      no     1136     yes     1  cellular   

      day month  duration  campaign  pdays  previous poutcome   y  \
0      19   oct        79         1     -1         0  unknown  no   
1      11   may       220         1    339         4  failure  no   
2      16   apr       185         1    330         1  failure  no   
3       3   jun       199         4     -1         0  unknown  no   
4       5   may       226         1     -1         0  unknown  no   
...   ...   ...       ...       ...    ...       ...      ...  ..   
4516   30   jul       329         5     -1         0  unknown  no   
4517    9   may       153         1     -1         0  unknown  no   
4518   19   aug       151        11     -1         0  unknown  no   
4519    6   feb       129         4    211         3    other  no   
4520    3   apr       345         2    249         7    other  no   

      marital_divorced  marital_married  marital_single  
0                    0                1               0  
1                    0                1               0  
2                    0                0               1  
3                    0                1               0  
4                    0                1               0  
...                ...              ...             ...  
4516                 0                1               0  
4517                 0                1               0  
4518                 0                1               0  
4519                 0                1               0  
4520                 0                0               1  

[4521 rows x 19 columns]

In [11]:
bank_data=pd.read_csv('/Users/sheelendersharma1/Desktop/data sciece/bank.csv',sep=';')
new_data=pd.get_dummies(bank_data,columns=['education','marital'])

new_data.head()

# if dont give prefix defaut the coloumn nmae it will take as prefix

age          job default  balance housing loan   contact  day month  \
0   30   unemployed      no     1787      no   no  cellular   19   oct   
1   33     services      no     4789     yes  yes  cellular   11   may   
2   35   management      no     1350     yes   no  cellular   16   apr   
3   30   management      no     1476     yes  yes   unknown    3   jun   
4   59  blue-collar      no        0     yes   no   unknown    5   may   

   duration  ...  previous  poutcome   y education_primary  \
0        79  ...         0   unknown  no                 1   
1       220  ...         4   failure  no                 0   
2       185  ...         1   failure  no                 0   
3       199  ...         0   unknown  no                 0   
4       226  ...         0   unknown  no                 0   

  education_secondary  education_tertiary  education_unknown  \
0                   0                   0                  0   
1                   1                   0                  0   
2                   0                   1                  0   
3                   0                   1                  0   
4                   1                   0                  0   

   marital_divorced  marital_married  marital_single  
0                 0                1               0  
1                 0                1               0  
2                 0                0               1  
3                 0                1               0  
4                 0                1               0  

[5 rows x 22 columns]

- curse of dimensionality


In [ ]:
## problem
# with this columns are incressing 
# assume that one feture having 20 lenels
# 20 extra columns are created
# id you have so much columns ----> curse of dimensionality


## Image analytics ---> 512*512
# 512*3 * 512*3--- huge
# Time consuming
# slow /gpu/TPUs

In [ ]:
-Method1: using map(many lables you can apply)
-Method2: np.where(only for binary lables)
    
-method3:pd.get_dumies(many labels)

### Method-4:


**Lable encoder**

- It will provide the lables automatically
- suppose you have two labels yes and no,then yes=0,no=1
- If you have 3 labels as single,married,divorce,thwn single=0,married=1,divorce=0
- Manual map is not required
- Supoose there are 100labels,automatically it will form 0 to 99
- best for ordinal values,ex: primary,secondary,teritory


https://scikit-learn.org/stable/

In [2]:
import sklearn

In [ ]:
# pip freeze ..

In [3]:
dir(sklearn)

['__SKLEARN_SETUP__',
 '__all__',
 '__builtins__',
 '__cached__',
 '__check_build',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_distributor_init',
 'base',
 'clone',
 'config_context',
 'exceptions',
 'externals',
 'get_config',
 'logger',
 'logging',
 'os',
 'random',
 'set_config',
 'setup_module',
 'show_versions',
 'sys',
 'utils']

In [4]:
# for preprocessing
from sklearn.preprocessing
# if you want to select important features
from sklearn.feature_selection
# if you want to select any models
from sklearn.model_selection

 # do not run this

SyntaxError: invalid syntax (1415993891.py, line 2)

In [ ]:
EDa----.>Exploratory data analysis
processing is one of technique in EDA

- step:1
    

In [5]:
from sklearn.preprocessing import LabelEncoder
# LablesEncoder is a class object
# class ---init----self----packages

In [6]:
sklearn  #<module 'sklearn' from '/Users/sheelendersharma1/opt/anaconda3/lib/python3.9/site-packages/sklearn/__init__.py'>


<module 'sklearn' from '/Users/sheelendersharma1/opt/anaconda3/lib/python3.9/site-packages/sklearn/__init__.py'>

In [ ]:
# whenever any class object you are reading
# first thing save that in one variable

- step:2

In [7]:
le=LabelEncoder()

In [8]:
le

LabelEncoder()

In [ ]:
# after storing class object in one variable
# you need perform the operation

# fit------transform---fit_transform
# what is fit:it is a calcutation

In [ ]:
**fit**
- for particular features it will calcuate mean and std
**Transform**
- you need to apply z-score formula on particular by using calculated mean and std
**Fit_transform:**
Both operation performed at atime

- step:3

In [12]:
bank_data['marital']=le.fit_transform(bank_data['marital'])

In [13]:
bank_data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed        1    primary      no     1787      no   no   
1   33     services        1  secondary      no     4789     yes  yes   
2   35   management        2   tertiary      no     1350     yes   no   
3   30   management        1   tertiary      no     1476     yes  yes   
4   59  blue-collar        1  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [ ]:
from sklearn.preprocessing import LabelEncoder # step1
le=LabelEncoder()#step2
bank_data['marital']=le.fit_transform(bank_data['marital'])#step3

SyntaxError: invalid syntax (1190819564.py, line 1)

#<img data-ils="4" jsaction="rcuQ6b:trigger.M8vzZb;" class="rg_i Q4LuWd" jsname="Q4LuWd" width="249" height="204" alt="Z-Score: Definition, Calculation &amp; Interpretation - Simply Psychology" src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTwdc-J3ixDPOPOhbNEHPfa5MkRjXSLWPQgHA&amp;usqp=CAU">

In [17]:
bank_data['marital'].values

array([1, 1, 2, ..., 1, 1, 2])

In [18]:
# apply inverse transform to get see the original labels
le.inverse_transform(bank_data['marital'])

array([1, 1, 2, ..., 1, 1, 2])

In [21]:
bank_data=pd.read_csv('/Users/sheelendersharma1/Desktop/data sciece/bank.csv',sep=';')

In [22]:
bank_data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [23]:
cati=bank_data.columns[bank_data.dtypes=='O']
cati

Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact',
       'month', 'poutcome', 'y'],
      dtype='object')

In [24]:
from sklearn.preprocessing import LabelEncoder  # step-1
le=LabelEncoder()   #step-2
bank_data['job']=le.fit_transform(data['job'])
bank_data['marital']=le.fit_transform(data['marital'])
bank_data['education']=le.fit_transform(data['education'])
bank_data['marital']=le.fit_transform(data['marital'])

# whenever you are doing same task multiple times think about for loop

NameError: name 'data' is not defined

In [25]:
for column in cati:
    bank_data[column]=le.fit_transform(bank_data[column])

In [26]:
import sys

In [27]:
sys.executable

'/Users/sheelendersharma1/opt/anaconda3/bin/python'